platforms/js/JsMd5.hx

```diff
@@ -0,0 +1,302 @@
+ */
+
+	function new(s : String) {
+		input = s;
+		inputBytes = 0;
+		nextCharIndex = 0;
+		bytesStorage = 0;
+		bytesInStorage = 0;
+		var i = 0;
+
+		while (i < s.length) {
+			var c : Int = StringTools.fastCodeAt(s, i++);
+			// surrogate pair
+			if (0xD800 <= c && c <= 0xDBFF) {
+				c = (c - 0xD7C0 << 10) | (StringTools.fastCodeAt(s, i++) & 0x3FF);
+			}
+			if (c <= 0x7F) {
+				inputBytes++;
+			} else if (c <= 0x7FF) {
+				inputBytes += 2;
+			} else if (c <= 0xFFFF) {
+				inputBytes += 3;
+			} else {
+				inputBytes += 4;
+			}
+		}
+	}
+
+	function bitOR(a, b) {
+		var lsb = (a & 0x1) | (b & 0x1);
+		var msb31 = (a >>> 1) | (b >>> 1);
+		return (msb31 << 1) | lsb;
+	}
+
+	function bitXOR(a, b) {
+		var lsb = (a & 0x1) ^ (b & 0x1);
+		var msb31 = (a >>> 1) ^ (b >>> 1);
+		return (msb31 << 1) | lsb;
+	}
+
+	function bitAND(a, b) {
+		var lsb = (a & 0x1) & (b & 0x1);
+		var msb31 = (a >>> 1) & (b >>> 1);
+		return (msb31 << 1) | lsb;
+	}
+
+	function addme(x, y) {
+		var lsw = (x & 0xFFFF)+(y & 0xFFFF);
+		var msw = (x >> 16)+(y >> 16)+(lsw >> 16);
+		return (msw << 16) | (lsw & 0xFFFF);
+	}
+
+	function hex(a : Array<Int>) {
+		var str = "";
+		var hex_chr = "0123456789abcdef";
+		for (num in a) {
+			for (j in 0...4) {
+				str += hex_chr.charAt((num >> (j * 8 + 4)) & 0x0F)
+						+ hex_chr.charAt((num >> (j * 8)) & 0x0F);
+			}
+		}
+		return str;
+	}
+
+	function rol(num, cnt) {
+		return (num << cnt) | (num >>> (32 - cnt));
+	}
+
+	function cmn(q, a, b, x, s, t) {
+		return addme(rol((addme(addme(a, q), addme(x, t))), s), b);
+	}
+
+	function ff(a, b, c, d, x, s, t) {
+		return cmn(bitOR(bitAND(b, c), bitAND((~b), d)), a, b, x, s, t);
+	}
+
+	function gg(a, b, c, d, x, s, t) {
+		return cmn(bitOR(bitAND(b, d), bitAND(c, (~d))), a, b, x, s, t);
+	}
+
+	function hh(a, b, c, d, x, s, t) {
+		return cmn(bitXOR(bitXOR(b, c), d), a, b, x, s, t);
+	}
+
+	function ii(a, b, c, d, x, s, t) {
+		return cmn(bitXOR(c, bitOR(b, (~d))), a, b, x, s, t);
+	}
+
+	// TODO: move to a separate class
+	function getNextByte() : Int {
+		if (bytesInStorage != 0) {
+			bytesInStorage--;
+			var result = bytesStorage & 0xFF;
+			bytesStorage >>= 8;
+			return result;
+		}
+
+		var result : Int = 0;
+		if (nextCharIndex >= input.length) { // Error
+			return result;
+		}
+
+		var c : Int = StringTools.fastCodeAt(input, nextCharIndex++);
+		// utf16-decode and utf8-encode
+		// surrogate pair
+		if (0xD800 <= c && c <= 0xDBFF ) {
+			c = (c - 0xD7C0 << 10) | (StringTools.fastCodeAt(input, nextCharIndex++) & 0x3FF);
+		}
+		
+		if (c <= 0x7F) {
+			result = c;
+		} else if (c <= 0x7FF) {
+			result = 0xC0 | (c >> 6);
+			bytesInStorage = 1;
+			bytesStorage = 0x80 | (c & 63);
+		} else if (c <= 0xFFFF) {
+			result = 0xE0 | (c >> 12);
+			bytesInStorage = 2;
+			bytesStorage = (0x80 | ((c >> 6) & 63)) | ((0x80 | (c & 63)) << 8);
+		} else {
+			result = 0xF0 | (c >> 18);
+			bytesInStorage = 3;
+			bytesStorage = (0x80 | ((c >> 12) & 63))
```

Comment:
we can replace 3 ```0x80 |``` with a single ```0x00808080 |```. But i decided not to introduce it right now - to make PR review easier.